### ISSUE TITLE
Mini Project: level 4 extra line of iterate

### ISSUE THREAD

**Issue by cindyangXQ** [nus-cs2030/2122-s1#317]
## Description 
Previous posts usually mentioned printing a extra line of `takeWhile: ` or `filter: `. However, I am printing out an extra line of `iterate: `. I thought the problem was with `iterate` so I changed its implementation, only to find out that in this way I would be missing a line of `iterate: ` in level 2&3 testcases. So the problem would be back to `limit`. Anyone had the same problem before can share how you solved it?

**Comment by Hongyi6328** [comment]
Usually this is because you eagerly evaluated something, ie, not being lazy enough.
For example, let's say a list consisting of `1, 2, 3, 4, 5, ...` is created via `list = InfiniteListImpl.iterate(1, x -> {System.out.println("iterate"); return x + 1;});`.
Applying `list = list.limit(4)`, the list now should be `1, 2, ,3, 4, EmptyList`.
If you evaluate the tail after `4` immediately after `4` is polled out using `peek()`, you actually triggered another `print()` method although the `EmptyList` wouldn't give you a `5`.
If my guess is right, I suggest you make you evaluation of the next tail lazy in the method `peek()`. Don't return a `tail` immediately, instead, return a "promise" of a `tail`.

**Comment by cindyangXQ** [comment]
Sorry I don't quite understand. `peek()` return a promise of the `head`? Isn't it supposed to return the `tail`?

**Comment by Hongyi6328** [comment]
> Sorry I don't quite understand. `peek()` return a promise of the `head`? Isn't it supposed to return the `tail`?

Ah sorry, I didn't make myself clear here, `peek()` indeed needs to return a new list consisting of the new `head` and the new `tail`, where the new `head` comes from the `head` of your current `tail`.
So I mean don't evaluate what is inside the new `head` in the method `peek()`

**Comment by cindyangXQ** [comment]
What I did is just `tail.get().get()`. Will this cause the evaluation of the head in the tail?

**Comment by Hongyi6328** [comment]
> What I did is just `tail.get().get()`. Will this cause the evaluation of the head in the tail?

I don't think so. Then is may be something else you didn't evaluated lazily enough. Try to check againt each line of code and think what still can be delayed later.
By the way, I don't think `Optional::get` is allowed for the mini project, try another method to circumvent this restriction.
